lib/Shell.js

```diff
@@ -0,0 +1,239 @@
+      throw err; // fallback
+    });
+
+    proc.once('close', code => {
+      psErr.end();
+      this.emit('end', code);
+      this.verbose && logger.info(`PS process ${this.pid} exited with code ${code}`);
+
+      if(!psErr.isEmpty() && isCriticalPSError(psErr.getContents())) {
+        throw new PS_PROC_ERROR(psErr.getContentsAsString());
+      }
+    });
+
+    this.invoke = this.invoke.bind(this, proc);
+    this.stop = this.stop.bind(this, proc);
+    this.dispose = this.dispose.bind(this, proc);
+
+    this.streams = {
+      stdin: proc.stdin,
+      stdout: proc.stdout,
+      stderr: proc.stderr,
+    };
+  }
+  addCommand(command = '', params = []) {
+    return new Promise(resolve => {
+      if(!(command instanceof PSCommand)) {
+        command = new PSCommand(command);
+      }
+
+      this.commands.push(command);
+
+      if(params && params.length > 0) {
+        logger.warn(`params argument for addCommand method will be deprecated soon. use addParameter or addParameters instead.\n
+        https://rannn505.gitbook.io/node-powershell/shell/addCommand`);
+        return this.addParameters(params);
+      }
+
+      return resolve(this.commands);
+    });
+  }
+  addArgument(argument = '') {
+    if(this.commands.length === 0) {
+      return Promise.reject(new PS_ARG_MISS_ERROR('Commands array is empty. please add at least one command before you use addArgument()'));
+    }
+    const lastCommand = this.commands.pop();
+    lastCommand.addArgument(argument);
+    return this.addCommand(lastCommand);
+  }
+  addParameter(parameter = {}) {
+    if(this.commands.length === 0) {
+      return Promise.reject(new PS_ARG_MISS_ERROR('Commands array is empty. please add at least one command before you use addParameter()'));
+    }
+    const lastCommand = this.commands.pop();
+    lastCommand.addParameter(parameter);
+    return this.addCommand(lastCommand);
+  }
+  addParameters(parameters = []) {
+    if(getType(parameters) !== 'Array') {
+      throw new PS_ARG_TYPE_ERROR('Parameters must be an array');
+    }
+    if(parameters.length === 0) {
+      throw new PS_ARG_MISS_ERROR('Parameters are missing');
+    }
+
+    return Promise.all(parameters.map(this.addParameter, this))
+      .then(() => Promise.resolve(this.commands));
+  }
+  clear() {
+    this.commands = [];
+    return Promise.resolve(this.commands);
+  }
+  invoke() {
+    return new Promise((resolve, reject) => {
+      this.invocationStateInfo = 'NotStarted'; // https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.psinvocationstate?view=powershellsdk-1.1.0
+      const EOI = shortid.generate();
+      const invocationCommands = this.commands.map(psCommand => psCommand.command).join('; ');
+      let invocationHadErrors = false;
+
+      const psOut = new ShellStreamBuffer(EOI);
+      this.streams.stdout.pipe(psOut);
+      this.streams.stderr.pipe(psOut);
+
+      this.streams.stderr.once('data', () => {
+        invocationHadErrors = true;
```

Comment:
Note that stderr having stuff written into it doesn't indicate an error - a non-zero status code does.

stderr although called literally "err" doesn't actually always signal errors - it signals metadata and diagnostic stuff you don't want in the output file when piping stuff. 

It is better to look at the return value for checking if the invocation had errors generally.

Reply:
lets talk about it :)